sonic_data_client/events_client.go

```diff
@@ -0,0 +1,368 @@
+const MISSED = "COUNTERS_EVENTS:missed_internal"
+const DROPPED = "COUNTERS_EVENTS:missed_by_slow_receiver"
+const LATENCY = "COUNTERS_EVENTS:latency_in_ms"
+
+var STATS_CUMULATIVE_KEYS = [...]string {MISSED, DROPPED}
+var STATS_ABSOLUTE_KEYS = [...]string {LATENCY}
+
+const STATS_FIELD_NAME = "value"
+
+const EVENTD_PUBLISHER_SOURCE = "{\"sonic-events-eventd"
+
+// Path parameter
+const PARAM_HEARTBEAT = "heartbeat"
+
+type EventClient struct {
+
+    prefix      *gnmipb.Path
+    path        *gnmipb.Path
+
+    q           *queue.PriorityQueue
+    channel     chan struct{}
+
+    wg          *sync.WaitGroup // wait for all sub go routines to finish
+
+    subs_handle unsafe.Pointer
+
+    stopped     int
+
+    // Stats counter
+    counters    map[string]uint64
+
+    last_latencies  [LATENCY_LIST_SIZE]uint64
+    last_latency_index  int
+    last_latency_full   bool
+
+    last_errors uint64
+}
+
+func set_heartbeat(val int) {
+    s := fmt.Sprintf("{\"HEARTBEAT_INTERVAL\":%d}", val)
+    rc := C.event_set_global_options(C.CString(s));
+    if rc != 0  {
+        log.V(4).Infof("Failed to set heartbeat val=%d rc=%d", val, rc)
+    }
+}
+
+func NewEventClient(paths []*gnmipb.Path, prefix *gnmipb.Path, logLevel int) (Client, error) {
+    var evtc EventClient
+    evtc.prefix = prefix
+    for _, path := range paths {
+        // Only one path is expected. Take the last if many
+        evtc.path = path
+    }
+
+    for _, e := range evtc.path.GetElem() {
+        keys := e.GetKey()
+        for k, v := range keys {
+            if (k == PARAM_HEARTBEAT) {
+                if val, err := strconv.Atoi(v); err == nil {
+                    log.V(7).Infof("evtc.heartbeat_interval is set to %d", val)
+                    set_heartbeat(val)
+                }
+            }
+        }
+    }
+
+    C.swssSetLogPriority(C.int(logLevel))
+
+    /* Init subscriber with cache use and defined time out */
+    evtc.subs_handle = C.events_init_subscriber_wrap(true, C.int(SUBSCRIBER_TIMEOUT))
+    evtc.stopped = 0
+
+    /* Init list & counters */
+    evtc.counters = make(map[string]uint64)
+
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for _, key := range STATS_ABSOLUTE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for i := 0; i < len(evtc.last_latencies); i++ {
+        evtc.last_latencies[i] = 0
+    }
+    evtc.last_latency_index = 0
+    evtc.last_errors = 0
+    evtc.last_latency_full = false
+
+    log.V(7).Infof("NewEventClient constructed. logLevel=%d", logLevel)
+
+    return &evtc, nil
+}
+
+
+func compute_latency(evtc *EventClient) {
+    if evtc.last_latency_full {
+        var total uint64 = 0
+        var cnt uint64 = 0
+
+        for _, v := range evtc.last_latencies {
+            if v > 0 {
+                total += v
+                cnt += 1
+            }
+        }
+        if (cnt > 0) {
+            evtc.counters[LATENCY] = (uint64) (total/cnt/1000/1000)
+        }
```

Comment:
when cnt is 0, what's the LATENCY value?

Reply:
cnt can't be 0. Just realized, as I was re-reading the code. Thanks for pointing out, which helped me re-analyze.